include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -2328,6 +2332,479 @@ namespace internal
+     *                   @p n_columns columns, stored in row-major format
+     * @param in Pointer to the start of the input data vector
+     * @param out Pointer to the start of the output data vector
+     */
+    template <int  direction,
+              bool contract_over_rows,
+              bool add,
+              int  normal_dir,
+              bool one_line = false>
+    static void
+    apply(const Number2 *DEAL_II_RESTRICT shape_data,
+          const Number *                  in,
+          Number *                        out);
+
+    template <int  face_direction,
+              bool contract_onto_face,
+              bool add,
+              int  max_derivative,
+              bool lex_faces = false,
+              int  normal_direction>
+    void
+    apply_face(const Number *DEAL_II_RESTRICT in,
+               Number *DEAL_II_RESTRICT       out) const;
+
+    const Number2 *shape_values;
+    const Number2 *shape_gradients;
+    const Number2 *shape_hessians;
+  };
+
+  template <int dim,
+            int n_rows,
+            int n_columns,
+            typename Number,
+            typename Number2>
+  template <int  direction,
+            bool contract_over_rows,
+            bool add,
+            int  normal_dir,
+            bool one_line>
+  inline void
+  EvaluatorTensorProduct<evaluate_raviart_thomas,
+                         dim,
+                         n_rows,
+                         n_columns,
+                         Number,
+                         Number2>::apply(const Number2 *DEAL_II_RESTRICT
+                                                       shape_data,
+                                         const Number *in,
+                                         Number *      out)
+  {
+    static_assert(one_line == false || direction == dim - 1,
+                  "Single-line evaluation only works for direction=dim-1.");
+    Assert(shape_data != nullptr,
+           ExcMessage(
+             "The given array shape_data must not be the null pointer!"));
+    Assert(dim == direction + 1 || one_line == true || n_rows == n_columns ||
+             in != out,
+           ExcMessage("In-place operation only supported for "
+                      "n_rows==n_columns or single-line interpolation"));
+    AssertIndexRange(direction, dim);
+    constexpr int mm = contract_over_rows ? n_rows : n_columns,
+                  nn = contract_over_rows ? n_columns : n_rows;
+
+    constexpr int stride    = Utilities::pow(n_columns, direction);
+    constexpr int n_blocks1 = one_line ? 1 : stride;
+
+    // The number of blocks depend on both direction and dimension.
+    constexpr int n_blocks2 =
+      (dim - direction - 1 == 0) ?
+        1 :
+        ((direction == normal_dir) ?
+           Utilities::pow((n_rows - 1),
+                          (direction >= dim) ? 0 : dim - direction - 1) :
+           (((direction < normal_dir) ? (n_rows + 1) : n_rows) *
```

Comment:
I remember discussing this with @jwitte08 two years ago :)

Reply:
I am wondering if we need two implementations of this function. Couldn't we extract the common parts?

Reply:
Yes, this is also mentioned by @kronbichler in #9545. I didn't want to do any major changes and potentially brake anything, but this should be done at some point.